### PR TITLE
Setup end-to-end tests

### PR DESCRIPTION
**Pull request by mhkc**
This PR adds end 2 end testing of the Bonsai stack using Selenium.

The PR also fixes a slew of bugs that were encountered when implementing the testing. Some of which are,
- Fixed issue where allele cluster service crashed if the allele profile were the same for all samples.
- Fixed issue that prevented multiple samples to be selected from the `/groups/<group_id>` view.
- Fixed issue where the frontend could crash if the sample did not have a cgmlst profile
- Added test samples.

Information on how the e2e tests work and can be run can be found in ./e2e_tests/README.md.

### CODE REVIEW

**Review by ryanjameskennedy — APPROVED** [review]
LGTM!

File: e2e_tests/tests/conftest.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    return element
+
+
+@pytest.fixture(scope="session")
+def remote_driver(config) -> webdriver:
+    """Setup remote driver."""
+    browsers = {
+        "chrome": ChromeOptions,
+        "firefox": FirefoxOptions,
+    }
```
Comment (alkc) [inline e2e_tests/tests/conftest.py:59]: i guess this sets up some sort of profile for selenium mimicking specific browsers.

if so, is microsoft edge something that can be added? dunno how tech-savvy the the end-users of bonsai are/will be but i guess most of them will use edge.

File: e2e_tests/tests/views/test_admin.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+from ..conftest import get_element_by_test_id
+
+def test_admin_console_not_accessable_by_user(logged_in_user, config):
+    """Test that the admin console is not accessable by user."""
+    # FIRST go to landing page
+    logged_in_user.get(str(Path(config["frontend_url"])))
+
+    # TEST that admin panel button is not accessable to a regular user
+    with pytest.raises(NoSuchElementException):
+        get_element_by_test_id(logged_in_user, "admin-panel-navbar-btn")
```
Comment (alkc) [inline e2e_tests/tests/views/test_admin.py:15]: nice!

File: e2e_tests/tests/views/test_cluster.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    (
+        ActionChains(logged_in_user)
+        .key_down(Keys.CONTROL)  # hold controll
+        .click(on_element = get_element_by_test_id(logged_in_user, f"sample-row-1"))  # select multiple samples
+        .click(on_element = get_element_by_test_id(logged_in_user, f"sample-row-2"))
+        .click(on_element = get_element_by_test_id(logged_in_user, f"sample-row-3"))
+        .click(on_element = get_element_by_test_id(logged_in_user, f"sample-row-4"))
+        .click(on_element = get_element_by_test_id(logged_in_user, f"sample-row-5"))
+        .key_up(Keys.CONTROL)
+    ).perform()
```
Comment (alkc) [inline e2e_tests/tests/views/test_cluster.py:51]: _nice._

File: e2e_tests/tests/views/test_group.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+@pytest.mark.parametrize("group_id", ["mtuberculosis", "saureus", "ecoli"])
+def test_open_group_view(logged_in_user, config, group_id: str):
+    """Test that the test groups are working."""
+
+    # FIRST goto the group view
+    url = Path(config["frontend_url"]) / 'groups' / group_id
+    logged_in_user.get(str(url))
+
+    # TEST that the page could load
+    assert "bonsai" in logged_in_user.title.lower()
```
Comment (alkc) [inline e2e_tests/tests/views/test_group.py:23]: is a http code == 200 test a good idea here?

**Review by alkc — APPROVED** [review]
Nice PR. Learned some new things from it!

I left a few comments, questions and suggestions but found no major issues.

Running:

```
docker compose -f docker-compose.yml -f docker-compose.e2e_test.yml run e2e-tests
```

Executes the end-to-end tests and generates a report as expected.

File: api/bonsai_api/cli.py [hunk 5]
```diff
@@ -80,24 +88,48 @@ def create_user(
...
+):  # pylint: disable=unused-argument
+    """Create a user account"""
+    # create collections
+    group_obj = GroupInCreate(group_id=id, display_name=name, description=description, 
+                              table_columns=pred_res_cols)
+    try:
+        loop = asyncio.get_event_loop()
+        with get_db_connection() as db:
+            func = create_group_in_db(db, group_obj)
+            loop.run_until_complete(func)
```
[the remark this answers is not in the document]
Comment (mhkc) [reply api/bonsai_api/cli.py:116]: Yes kind of like that. The API initilaize the db connection at each requests which minimizes the risk of timing issues (the api being up before the database is ready to take connections). This is a workaround so I can use the same code for the cli.

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        branches: ["master"]
+    pull_request:
+        branches: ["master"]
+
+jobs:
+  unit-tests:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python-version: ["3.11", "3.12"]
```
[the remark this answers is not in the document]
Comment (mhkc) [reply .github/workflows/ci.yml:16]: Its for now. We should bump python in the containers and drop 3.11

File: e2e_tests/tests/conftest.py (hunk 1 above)
Comment (mhkc) [reply e2e_tests/tests/conftest.py:59]: Yes we can run edge as well. I currently only run chrome beacause that was the webdriver I had previously set up. But in the future I wanted to setup so all tests are run with chrome, edge and firefox but this requires the setup of a selenium grid and was a little bit out of scope for this.

Comment (alkc) [reply e2e_tests/tests/conftest.py:59]: gotcha

File: e2e_tests/tests/views/test_group.py (hunk 4 above)
Comment (mhkc) [reply e2e_tests/tests/views/test_group.py:23]: Selenium doesnt expose the HTTP response code by default. People have been creating workaround like this for that purpose, https://gist.github.com/LeMoussel/bfed69044cb65947a4e63c76f11e6e1f.

Comment (mhkc) [reply e2e_tests/tests/views/test_group.py:23]: I think that is something we could add in a followup PR perhaps as the tests grows.

**Pull request merged**